### PR TITLE
Implement Service to Create Opensearch Indexes of Artifacts 

### PR DESCRIPTION
**Pull request by puru-khedre**
I've Implemented the below services
### 1. `create#ArtifactLogIndexs`: (wrapper service)
This service lists all the newly updated artifact logs and calls the core service to create indexes of artifacts

### 2. `create#ArtifactLogIndex`: (core service)
Core service to create an index of a single artifact log, this service handles the content based on the content file format

### CODE REVIEW

File: service/co/hotwax/alc/ArtifactLifeCycleServices.xml [hunk 1]
```diff
@@ -1,3 +1,126 @@
...
+                <entity-make-value entity-name="moqui.service.job.ServiceJobParameter" value-field="updateLastRunParam" map="lastRunParam"/>
+                <set field="updateLastRunParam.parameterValue" from="nowDate"/>
+                <entity-update value-field="updateLastRunParam"/>
+            </if>
+
+            <return message="All Artifact Logs are indexed that are updated after Datetime: ${sinceDate}"/>
+        </actions>
+    </service>
+    <service verb="create" noun="ArtifactLogIndex">
+        <description>This service processes the provided artifact log and creates an index on opensearch.</description>
```
Comment (dt2patel) [inline service/co/hotwax/alc/ArtifactLifeCycleServices.xml]:
```suggestion
        <description>This service processes the provided Artifact Log and creates an index on OpenSearch.</description>
```

File: service/co/hotwax/alc/ArtifactLifeCycleServices.xml [hunk 2]
```diff
@@ -1,3 +1,126 @@
...
+        <in-parameters>
+            <parameter name="artifactEventId"/>
+            <parameter name="artifactLogValue" type="Map"/>
```
Comment (dt2patel) [inline service/co/hotwax/alc/ArtifactLifeCycleServices.xml]: This service should only accept event Id not a map. event ID is required

File: service/co/hotwax/alc/ArtifactLifeCycleServices.xml [hunk 3]
```diff
@@ -1,3 +1,126 @@
...
+
+                        for (Map.Entry<String, Integer> field in headerMap) {
+                            fieldValue = record.get(field.value)
+                            if (fieldValue == null) continue
+                            valueMap.put(field.key, fieldValue.isEmpty() ? null : fieldValue)
+                        }
+                        contentObject.add(valueMap)
+                    }
+                } else {
+                    ec.message.addError("File Format ${contentType} is not Supported")
```
Comment (dt2patel) [inline service/co/hotwax/alc/ArtifactLifeCycleServices.xml]:
```suggestion
                    ec.message.addError("${contentType} file format is not supported. Please upload a JSON or CSV")
```

**Review by dt2patel — CHANGES_REQUESTED** [review]
See suggested commits as well as revise based on design discussion we had in respect to how the core indexing service should function.

**Comment by puru-khedre** [comment]
@dt2patel sir, We can merge this PR now all comments have been resolved

Creating a separate PR to Implement the Artifact Logs create service

File: service/co/hotwax/alc/ArtifactLifeCycleServices.xml [hunk 4]
```diff
@@ -1,3 +1,126 @@
...
+            <set field="nowDate" from="ec.user.nowTimestamp"/>
+
+            <entity-find entity-name="co.hotwax.alc.ArtifactLog" list="artifactLogList">
+                <econdition field-name="lastUpdatedStamp" from="sinceDate" operator="greater" ignore-if-empty="true"/>
+                <select-field field-name="artifactEventId"/>
+            </entity-find>
+
+            <!-- Iterate through the list of artifact logs and process each one -->
+            <set field="createIndexCount" value="0" type="Integer"/>
+            <iterate list="artifactLogList" entry="artifactLogValue">
```
Comment (dt2patel) [inline service/co/hotwax/alc/ArtifactLifeCycleServices.xml]:
```suggestion
            <iterate list="artifactLogList" entry="artifactEvent">
```

File: service/co/hotwax/alc/ArtifactLifeCycleServices.xml [hunk 5]
```diff
@@ -1,3 +1,126 @@
...
+                <service-call name="co.hotwax.alc.ArtifactLifeCycleServices.create#ArtifactLogIndex" in-map="[artifactEventId:artifactLogValue.artifactEventId]"/>
```
Comment (dt2patel) [inline service/co/hotwax/alc/ArtifactLifeCycleServices.xml]:
```suggestion
                <service-call name="co.hotwax.alc.ArtifactLifeCycleServices.create#ArtifactLogIndex" in-map="[artifactEventId:artifactEvent.artifactEventId]"/>
```

File: service/co/hotwax/alc/ArtifactLifeCycleServices.xml [hunk 6]
```diff
@@ -1,3 +1,126 @@
...
+
+            <!-- Update the last run time of the job -->
+            <if condition="jobName">
+                <service-call
+                    name="update#moqui.service.job.ServiceJobParameter"
+                    in-map="[jobName:jobName, parameterName:'lastRunTime', parameterValue:nowDate]"
+                />
+            </if>
+
+            <return message="Total ${createIndexCount} Artifact Logs are indexed that are updated after Datetime: ${sinceDate}"/>
```
Comment (dt2patel) [inline service/co/hotwax/alc/ArtifactLifeCycleServices.xml]:
```suggestion
            <return message="${createIndexCount} Artifact Logs were indexed from ${sinceDate}"/>
```

File: service/co/hotwax/alc/ArtifactLifeCycleServices.xml [hunk 7]
```diff
@@ -1,3 +1,126 @@
...
+        </actions>
+    </service>
+    <service verb="create" noun="ArtifactLogIndex">
+        <description>This service processes the provided Artifact Log and creates an index on OpenSearch.</description>
+        <in-parameters>
+            <parameter name="artifactEventId" required="true"/>
+        </in-parameters>
+        <actions>
+            <!-- Find the corresponding artifact log value -->
+            <entity-find-one entity-name="co.hotwax.alc.ArtifactLog" value-field="artifactLogValue"/>
```
Comment (dt2patel) [inline service/co/hotwax/alc/ArtifactLifeCycleServices.xml]:
```suggestion
            <entity-find-one entity-name="co.hotwax.alc.ArtifactLog" value-field="artifactLog"/>
```

File: service/co/hotwax/alc/ArtifactLifeCycleServices.xml [hunk 8]
```diff
@@ -1,3 +1,126 @@
...
+            <if condition="artifactLogValue == null"><return error="true" message="Artifact Log is not found"/></if>
```
Comment (dt2patel) [inline service/co/hotwax/alc/ArtifactLifeCycleServices.xml]:
```suggestion
            <if condition="artifactLog == null"><return error="true" message="Artifact Log is not found"/></if>
```

**Review by dt2patel — CHANGES_REQUESTED** [review]
Update naming

File: data/ArtifactLogServiceJobData.xml [hunk 9]
```diff
@@ -0,0 +1,14 @@
+<?xml version="1.0" encoding="UTF-8"?>
+
+<entity-facade-xml type="ext">
+    <moqui.service.job.ServiceJob
+            jobName="create_ArtifactLogIndexes"
+            description="consume all of the newly created Artifact Logs"
```
Comment (dt2patel) [inline data/ArtifactLogServiceJobData.xml]:
```suggestion
            description="Index recently updated Artifact Logs"
```

File: service/co/hotwax/alc/ArtifactLifeCycleServices.xml [hunk 10]
```diff
@@ -1,3 +1,126 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <services xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="http://moqui.org/xsd/service-definition-3.xsd">
+    <service verb="create" noun="ArtifactLogIndexes">
+        <description>This service indexes each artifact log by calling the core service. It only process logs after the 'sinceDate', If not specified, it defaults to the last run time of the job.</description>
+        <in-parameters>
+            <parameter name="sinceDate" type="Timestamp">
+                <description>The timestamp to filter artifact logs updated after this date</description>
```
Comment (dt2patel) [inline service/co/hotwax/alc/ArtifactLifeCycleServices.xml]:
```suggestion
                <description>Filter Artifact Logs updated after this date time</description>
```

**Review by dt2patel — CHANGES_REQUESTED** [review]

**Review by dt2patel — APPROVED** [review]

**Pull request merged**